### PR TITLE
 implement TripletMarginLoss as a native function

### PR DESCRIPTION
**Pull request by li-roy**
- Implement TripletMarginLoss as a native function
- add size_average arg
- add reduce arg as per #264 

Benchmarks run on inputs of size (5,500) x 5000 iterations.

old:
forward [1.0334448860958219, 1.0698160571046174, 1.054812144022435]
backward [1.6976923400070518, 1.6186234278138727, 1.617640192154795]
double backward [3.2021269120741636, 3.165953448973596, 3.24088541790843]

new:
forward [0.8313469190616161, 0.8255189431365579, 0.8157028129789978]
backward [1.5887232730165124, 1.6056769208516926, 1.5178304109722376]
double backward [3.154403690015897, 3.1905114939436316, 3.2450901730917394]

### CODE REVIEW

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -959,16 +967,17 @@ class TripletMarginLoss(Module):
         http://www.iis.ee.ic.ac.uk/%7Evbalnt/shallow_descr/TFeat_paper.pdf
     """
 
-    def __init__(self, margin=1.0, p=2, eps=1e-6, swap=False):
-        super(TripletMarginLoss, self).__init__()
+    def __init__(self, margin=1.0, p=2, eps=1e-6, swap=False, size_average=True, reduce=True):
```
Comment (zou3519) [inline torch/nn/modules/loss.py]: It would be nice to document `swap`

File: test/test_nn.py [hunk 2]
```diff
@@ -4844,6 +4866,17 @@ def forward(self, *args):
         check_no_size_average=True,
         check_gradgrad=False,
     ),
+    dict(
+        module_name='TripletMarginLoss',
+        constructor_args=(torch.rand(10),),
+        input_fn=lambda: torch.randn(5, 10),
+        target_fn=lambda: torch.rand(5, 10).mul(2).floor(),
+        reference_fn=lambda i, t, m: -((t * i.sigmoid().log() + (1 - t) * (-i).sigmoid().log()) * get_weight(m)).sum() /
```
Comment (zou3519) [inline test/test_nn.py]: Is this the formulation of TripletMarginLoss?

File: aten/src/ATen/native/Distance.cpp [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+
+
+namespace at { namespace native {
+
+Tensor pairwise_distance(const Tensor& x1, const Tensor& x2, double p, double eps) {
+  auto diff = abs(x1 - x2);
+  auto out = pow(diff + eps, p).sum(1);
+  return pow(out, 1 / p);
```
Comment (zou3519) [inline aten/src/ATen/native/Distance.cpp]: Isn't this just `tensor.norm`?

File: torch/nn/modules/loss.py [hunk 4]
```diff
@@ -933,20 +933,31 @@ class TripletMarginLoss(Module):
...
 
     Args:
-        anchor: anchor input tensor
-        positive: positive input tensor
-        negative: negative input tensor
-        p: the norm degree. Default: 2
+        margin (float, optional): Default: `1`.
+        p (int, optional): The norm degree for pairwise distance. Default: `2`.
+        swap (float, optional): The distance swap is described in detail in the paper
+            `Learning shallow convolutional feature descriptors with triplet losses` by
```
Comment (zou3519) [inline torch/nn/modules/loss.py]: Nit: does this actually link to the paper in the link below? You can check by rebuilding the docs (cd into docs/, then `make html`). The original string has a _ after the ending ` so I was wondering if that did anything.

File: torch/nn/modules/loss.py (hunk 1 above)
Comment (zou3519) [reply torch/nn/modules/loss.py]: Documented.

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -271,6 +271,9 @@
 - func: ones_like(Tensor self, *, Type dtype) -> Tensor
   variants: function
 
+- func: pairwise_distance(Tensor x1, Tensor x2, double p, double eps) -> Tensor
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml]: You can (and probably should, to match the python API) set a default value for eps and p.

File: aten/src/ATen/native/native_functions.yaml [hunk 6]
```diff
@@ -360,6 +363,9 @@
 - func: transpose_(Tensor self, int64_t dim0, int64_t dim1) -> Tensor
   variants: method
 
+- func: triplet_margin_loss(Tensor anchor, Tensor positive, Tensor negative, double margin, double p, double eps, bool swap, bool size_average, bool reduce) -> Tensor
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml]: You can set default values for margin, p, eps, swap, size_average, and reduce

File: aten/src/ATen/native/Distance.cpp [hunk 7]
```diff
@@ -0,0 +1,11 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+
+
+namespace at { namespace native {
+
+Tensor pairwise_distance(const Tensor& x1, const Tensor& x2, double p, double eps) {
+  auto diff = abs(x1 - x2 + eps);
+  return norm(diff, p, 1);
```
Comment (zou3519) [inline aten/src/ATen/native/Distance.cpp]: Nit: norm already takes the abs of it's input, so you can rewrite pairwise_distance as `return norm(x1 - x2 + eps, p, 1)`

File: aten/src/ATen/native/Distance.cpp [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+
+
+namespace at { namespace native {
+
+Tensor pairwise_distance(const Tensor& x1, const Tensor& x2, double p, double eps) {
+  auto diff = abs(x1 - x2 + eps);
```
Comment (zou3519) [inline aten/src/ATen/native/Distance.cpp]: I don't really understand why `eps` is necessary, but it was in the original code... do you know why?

File: torch/nn/functional.py [hunk 9]
```diff
@@ -1960,11 +1960,7 @@ def pairwise_distance(x1, x2, p=2, eps=1e-6):
         >>> output = F.pairwise_distance(input1, input2, p=2)
         >>> output.backward()
     """
-    assert x1.size() == x2.size(), "Input sizes must be equal."
-    assert x1.dim() == 2, "Input must be a 2D matrix."
-    diff = torch.abs(x1 - x2)
-    out = torch.pow(diff + eps, p).sum(dim=1, keepdim=True)
```
Comment (zou3519) [inline torch/nn/functional.py]: I'm a little worried because it looks like we're changing the behavior of `pairwise_distance` to return a tensor of size `(N,)` instead of size `(N, 1)`.

I think keeping the old behavior (return tensor of size `(N, 1)`) might be better because it represents batched distances. On the other hand, this makes it inconsistent with our loss functions (but I guess `pairwise_distance` isn't a loss, and you can keep the current behavior of `triplet_margin_loss`. What do you think?

File: torch/nn/functional.py [hunk 10]
```diff
@@ -1960,11 +1960,7 @@ def pairwise_distance(x1, x2, p=2, eps=1e-6):
         >>> output = F.pairwise_distance(input1, input2, p=2)
```
Comment (zou3519) [inline torch/nn/functional.py]: GitHub won't let me tag the like I want, but... in the docs for pairwise_distance you should change the output shape to `(N,)`.

**Review by zou3519 — COMMENTED** [review]
Code looks good to me. I had a few minor comments and a concern that we're subtly changing the behavior of `pairwise_distance`

File: torch/nn/functional.py (hunk 9 above)
Comment (li-roy) [reply torch/nn/functional.py]: The reason I did this was because legacy seems to return (N) size. https://github.com/pytorch/pytorch/blob/master/torch/legacy/nn/PairwiseDistance.py#L30

I'm not sure why the change was made, but I think we should make them the same. (By changing legacy behavior to (N, 1) maybe?)

Comment (zou3519) [reply torch/nn/functional.py]: If any users use `pairwise_distance` right now (the non-legacy version) we'll be breaking their code. Not sure if that's a good thing or a bad thing. We should definitely keep the legacy code the same, it's there for interfacing with users of lua torch.

File: aten/src/ATen/native/Distance.cpp [hunk 11]
```diff
@@ -5,7 +5,6 @@
 namespace at { namespace native {
 
 Tensor pairwise_distance(const Tensor& x1, const Tensor& x2, double p, double eps) {
```
Comment (zou3519) [inline aten/src/ATen/native/Distance.cpp]: nit: Can we have this (and the public API for pairwise_distance) take in a `keepdim` option like the rest of our reduction functions, so that if we do break any user's code, it'll be a one-keyword fix?

**Review by zou3519 — APPROVED** [review]
LGTM. Minor nit below

File: aten/src/ATen/native/Distance.cpp (hunk 11 above)
Comment (li-roy) [reply aten/src/ATen/native/Distance.cpp]: Isn't it already a super simple fix? (squeeze) Is this worth it?

Comment (zou3519) [reply aten/src/ATen/native/Distance.cpp]: I think it's worth it to be consistent with the rest of our reduction API. I'm not sure what people use pairwise_distance for in real life though.

**Review by zou3519 — APPROVED** [review]
Yay! Approve approve approve

File: torch/nn/functional.py (hunk 9 above)
Comment (ahirner) [reply torch/nn/functional.py]: Pairwise distance takes the distance for each element to each element in another set (NxM) (http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.pairwise_distances.html).
This function calculates one distance between two vectors (N).

Since it's a relatively fresh PR I'm asking here: @li-roy can we have some clarification?
